homeassistant/components/media_player/template_media_player.py

```diff
@@ -0,0 +1,194 @@
+from homeassistant.const import CONF_NAME, STATE_ON, STATE_OFF
+from homeassistant.helpers.script import Script
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_ON_ACTION = 'turn_on_action'
+CONF_OFF_ACTION = 'turn_off_action'
+CONF_VOLUME_UP_ACTION = 'volume_up_action'
+CONF_VOLUME_DOWN_ACTION = 'volume_down_action'
+CONF_INITIAL_VOLUME = 'initial_volume'
+CONF_VOLUME_STEP = 'volume_step'
+CONF_VOLUME_STEP_DELAY = 'volume_step_delay'
+CONF_VOLUME_SET_ACTION = 'volume_set_action'
+CONF_VOLUME_MUTE_ACTION = 'volume_mute_action'
+CONF_VOLUME_UNMUTE_ACTION = 'volume_unmute_action'
+CONF_SOURCE_NAMES = 'source_names'
+CONF_SELECT_SOURCE_ACTION = 'select_source_action'
+DEFAULT_NAME = "Template Media Player"
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_ON_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_OFF_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_VOLUME_UP_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_VOLUME_DOWN_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_INITIAL_VOLUME, default=50): cv.positive_int,
+    vol.Optional(CONF_VOLUME_STEP, default=5): cv.positive_int,
+    vol.Optional(CONF_VOLUME_STEP_DELAY, default=0.5): cv.positive_int,
+    vol.Optional(CONF_VOLUME_SET_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_VOLUME_MUTE_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_VOLUME_UNMUTE_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_SELECT_SOURCE_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_SOURCE_NAMES, default=None): {cv.string: cv.string},
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the template media player platform."""
+    media_player = TemplateMediaPlayerDevice(config.get(CONF_NAME), config,
+                                             hass)
+    add_devices([media_player])
+
+
+class TemplateMediaPlayerDevice(MediaPlayerDevice):
+    """ Represents a generic device that delegates all operations to other services """
+
+    def __init__(self, name, config, hass):
+        """Initialize the device."""
+        self._name = name
+        self._turn_on_action = Script(hass, config.get(
+            CONF_ON_ACTION)) if config.get(CONF_ON_ACTION) else None
+        self._turn_off_action = Script(hass, config.get(
+            CONF_OFF_ACTION)) if config.get(CONF_OFF_ACTION) else None
+        self._volume_up_action = Script(hass, config.get(
+            CONF_VOLUME_UP_ACTION)) if config.get(
+            CONF_VOLUME_UP_ACTION) else None
+        self._volume_down_action = Script(hass, config.get(
+            CONF_VOLUME_DOWN_ACTION)) if config.get(
+            CONF_VOLUME_DOWN_ACTION) else None
+        self._volume_set_action = Script(hass, config.get(
+            CONF_VOLUME_SET_ACTION)) if config.get(
+            CONF_VOLUME_SET_ACTION) else None
+        self._volume_mute_action = Script(hass, config.get(
```

Comment:
Multi-line if's are very unreadable.

I think it would make more sense to just convert all of them to a script and store them as a dict:

```python
self._actions = {key: Script(hass, value) for key, value in config.items()
                 if key.endswith('_action')}
```

In your methods you can then do something like:

```python
if CONF_VOLUME_SET_ACTION in self._actions:
    yield from self._actions[CONF_VOLUME_SET_ACTION].async_run()
```